### PR TITLE
[feat] Track entropy and MI of routing distribution for topk MoE

### PR DESCRIPTION
**Pull request by oleksost**
# ✨ Description
To better detect potential routing collapse and have a better understanding about the routing distribution, we can track the average entropy and mutual information (which is the difference between the entropy of average routing and average routing entropy) of routing probabilities.

Collapse routing would have low entropy and low mutual information. A healthy and specialised router would have low entropy and high mutual information. 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes
- added calculation of both metrics in the `mixture_of_experts.py`, they are calculated only for the topk routing type. 

### General

- [x] 📜 I have read and followed the [contributing guidelines](https://servicenow.github.io/Fast-LLM/developers/contributing).
- [x] 🏷️ I am using a clear and descriptive PR title that summarizes the key change or feature introduced.
- [x] 🎉 The functionality is complete, and I have tested the changes.
- [x] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [x] 🧩 I have commented my code, especially in hard-to-understand areas.

### Testing

- [x] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [x] 🚦 I have tested these changes on GPUs and verified training stability.
- [x] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

## 📊 Performance Impact Details

I am not 100% sure there is no performance impact, we are calculating the stats at each forward pass through the router.

---

## 🗒️ Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

**Review by tscholak — COMMENTED** [review]
idea is good, thanks @oleksost.
bit weird that all these metrics are appearing as losses. that name should be reserved for things for which gradients are computed. just call this dict metrics?

**Comment by oleksost** [comment]
Yes, indeed, this has to be improved.

File: fast_llm/engine/base_model/base_model.py [hunk 1]
```diff
@@ -135,3 +135,7 @@ def get_tied_weights(self) -> dict[str, tuple[ParameterMeta, tuple[int, ...]]]:
     @abc.abstractmethod
     def loss_defs(self) -> list[LossDef]:
         pass
+
+    @property
```
Comment (jlamypoirier) [inline fast_llm/engine/base_model/base_model.py:139]: This loss/metric split is way more complicated than needed. How about having a single entry, and using a `is_metric` flag in `LossDef` (or a derived class) to distinguish? Then no change is needed other than extracting metrics from the context before returning from `run_step`

File: fast_llm/engine/schedule/runner.py [hunk 2]
```diff
@@ -289,6 +312,19 @@ def _reduce_losses(self, context: BatchContext) -> dict[str, float | int]:
             for name, reduced_loss in reduced_losses.items()
         }
 
+    def _is_reduced_metric(self, metric_name: str) -> bool:
+        """Check if a metric should be reduced (is defined in a TransformerReducedMetrics subclass)."""
+        from fast_llm.layers.transformer.config import TransformerReducedMetrics
```
Comment (jlamypoirier) [inline fast_llm/engine/schedule/runner.py:317]: We can't use hard-coded values here. Suggestion above would fix it, or there are a few other ways to get this dynamically.

File: fast_llm/layers/transformer/mixture_of_experts.py [hunk 3]
```diff
@@ -169,11 +199,27 @@ def _topk_routing(
...
+        metrics: dict | None = None,
     ) -> tuple[torch.Tensor, torch.Tensor]:
         top_logits, top_experts = torch.topk(logits, k=self._experts_per_token, dim=-1)
         scores = torch.softmax(top_logits, dim=-1, dtype=torch.float32)
         if losses is not None or (self.training and grad_scale is not None):
             probs = torch.softmax(logits, dim=-1, dtype=torch.float32)
+
+
+            # Store these metrics        
+            if metrics is not None:
```
Comment (jlamypoirier) [inline fast_llm/layers/transformer/mixture_of_experts.py:211]: Given the extra computation involved, this should be enabled through a config parameter

File: tests/test_moe_metrics.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                [0.1, 0.7, 0.1, 0.1],  # Token 3: mostly expert 1
+            ],
+        ]
+    )
+    mutual_info = calculate_mutual_information(mixed_probs)
+    # The expected value is between 0 and 1
+    assert 0.0 < mutual_info < 1.0, f"Expected value between 0 and 1, got {mutual_info}"
+
+
+def test_edge_cases():
```
Comment (jlamypoirier) [inline tests/test_moe_metrics.py:157]: More explicit name?

File: tests/test_moe_metrics.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    small_probs[:, :, 0] = 1.0 - 3e-8  # Make sure they sum to 1
+    norm_entropy = calculate_normalized_average_entropy(small_probs)
+    mutual_info = calculate_mutual_information(small_probs)
+    assert torch.isclose(norm_entropy, torch.tensor(0.0), atol=1e-5), f"Expected ~0.0, got {norm_entropy}"
+    assert torch.isclose(mutual_info, torch.tensor(0.0), atol=1e-5), f"Expected ~0.0, got {mutual_info}"
+
+
+
+@pytest.fixture
+def setup_runner():
```
Comment (jlamypoirier) [inline tests/test_moe_metrics.py:177]: These don't belong here. How about `test_runner.py`?

File: tests/test_moe_metrics.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    
+    reduced_losses = runner._reduce_losses(context)
+    
+    assert "test_loss" in reduced_losses
+    assert pytest.approx(reduced_losses["test_loss"], 0.001) == 2.0
+
+
+
+
+if __name__ == "__main__":
```
Comment (jlamypoirier) [inline tests/test_moe_metrics.py:281]: Not needed

File: fast_llm/layers/transformer/mixture_of_experts.py [hunk 7]
```diff
@@ -26,6 +27,35 @@
 logger = logging.getLogger(__name__)
 
 
+def calculate_normalized_average_entropy(probs: torch.Tensor) -> torch.Tensor:
```
Comment (jlamypoirier) [inline fast_llm/layers/transformer/mixture_of_experts.py:30]: Could try `@torch.compile` on these for a free performance boost.

File: fast_llm/layers/transformer/mixture_of_experts.py [hunk 8]
```diff
@@ -26,6 +27,35 @@
...
+    """
+    Calculates routing entropy for each token, then averages over all tokens.
+    If low, means a lot of mass is put on a single expert in all tokens, which can indicate collapse or specialization.
+    """
+    n_experts = probs.size(-1)
+    entropy_values = entropy(probs)
+    average_entropy = entropy_values.mean()  # Average over batch and tokens
+    return average_entropy / torch.log(torch.tensor(n_experts, dtype=probs.dtype, device=probs.device))
+
+def entropy(probs: torch.Tensor) -> torch.Tensor:
```
Comment (jlamypoirier) [inline fast_llm/layers/transformer/mixture_of_experts.py:40]: `calculate_entropy`

**Review by jlamypoirier — COMMENTED** [review]
Looks good, got some comments on the structure.

File: tests/test_moe_metrics.py (hunk 5 above)
Comment (tscholak) [reply tests/test_moe_metrics.py:177]: why don't they belong here? this is fixture is only useful for the tests in this suite.

File: fast_llm/layers/transformer/mixture_of_experts.py (hunk 3 above)
Comment (tscholak) [reply fast_llm/layers/transformer/mixture_of_experts.py:211]: how much compute are we talking about for these metrics? likely this won't be noticeable.

**Comment by jlamypoirier** [comment]
@oleksost Are you still working on this?

**Comment by oleksost** [comment]
@jlamypoirier yes, will address your comments today. Sorry, it was prioritised in favour of mamba.

File: tests/test_moe_metrics.py (hunk 5 above)
Comment (oleksost) [reply tests/test_moe_metrics.py:177]: Maybe we can move it to common.py in the future, it maybe be reused by other tests (e.g. ssm test)

File: fast_llm/layers/transformer/mixture_of_experts.py (hunk 3 above)
Comment (oleksost) [reply fast_llm/layers/transformer/mixture_of_experts.py:211]: this is already controlled by `training.logs.interval` parameter afaiu. Do you think we need a seperate parameter for MoE stats logging?

File: fast_llm/engine/schedule/runner.py (hunk 2 above)
Comment (oleksost) [reply fast_llm/engine/schedule/runner.py]: Simplified the setup s.t. all metrics that come back from a forward pass are reduced automatically, hence no need for this function.

File: fast_llm/engine/base_model/base_model.py (hunk 1 above)
Comment (oleksost) [reply fast_llm/engine/base_model/base_model.py:139]: This would be nice! 

Maybe better to leave it for a separate pr? It would make this one larger as it would require also changing the interfaces of the models' forward functions (that expect losses and metrics) as well as making sure that metrics are only calculated when `return_metrics` is True.

File: tests/test_moe_metrics.py (hunk 5 above)
Comment (jlamypoirier) [reply tests/test_moe_metrics.py:175]: I'm talking about the associated tests, not the fixture. They test a feature of `ScheduleRunner` and have nothing to do with MoE other than the loss names which aren't relevant to the tests.

File: fast_llm/layers/transformer/mixture_of_experts.py (hunk 3 above)
Comment (jlamypoirier) [reply fast_llm/layers/transformer/mixture_of_experts.py:218]: Ideally we'd like to calculate the bare minimum by default (and the computation isn't optimized), so I think yes.

If we really want by default I guess we could have a parameter with that defaults to true but can disabled for performance, eg. for benchmarks.

Comment (oleksost) [reply fast_llm/layers/transformer/mixture_of_experts.py]: addressed

**Comment by oleksost** [comment]
@jlamypoirier I think I addressed all the comments.

File: fast_llm/engine/base_model/base_model.py (hunk 1 above)
Comment (jlamypoirier) [reply fast_llm/engine/base_model/base_model.py:139]: There isn't much change needed actually, just need to add `kwargs["return_metrics"]`. I would prefer doing this here so we don't grow `ScheduleRunner` too much.
